### PR TITLE
Bugfix/unwraps

### PR DESCRIPTION
**Pull request by gpwclark**

### CODE REVIEW

File: src/color.rs [hunk 1]
```diff
@@ -316,19 +316,19 @@ fn detect_color(conin: &mut dyn ConsoleRead, color: u16) -> io::Result<bool> {
...
-            Ok(b) => total_read += b,
-            // Don't error out on a would block- this just means no response since async.
-            Err(err) if err.kind() == io::ErrorKind::WouldBlock => {}
-            Err(err) => return Err(err),
+    if conin.poll_timeout(timeout) {
+        while buf[0] != bell {
+            match conin.read(&mut buf) {
+                Ok(b) => total_read += b,
+                // Don't error out on a would block- this just means no response since async.
+                Err(err) if err.kind() == io::ErrorKind::WouldBlock => {}
```
Comment (sstanfield) [inline src/color.rs]: I think this line (and it's comment) should go away.  Now that you are polling conin then this should not happen.  Maybe it could with some sort of interrupt to the underlying select (but I don't think poll will return true then anyway) but even so I think just falling through to the general error is better.  If for some unknown reason this did happen we might wind up blocking when the while loop repeat and calls read without the poll.

File: src/color.rs [hunk 2]
```diff
@@ -316,19 +316,19 @@ fn detect_color(conin: &mut dyn ConsoleRead, color: u16) -> io::Result<bool> {
...
 
     // Either consume all data up to bell or wait for a timeout.
-    while buf[0] != bell && now.elapsed().unwrap() < timeout {
-        match conin.read(&mut buf) {
-            Ok(b) => total_read += b,
-            // Don't error out on a would block- this just means no response since async.
-            Err(err) if err.kind() == io::ErrorKind::WouldBlock => {}
-            Err(err) => return Err(err),
+    if conin.poll_timeout(timeout) {
+        while buf[0] != bell {
```
Comment (sstanfield) [inline src/color.rs:324]: I think this while loop being inside the poll is a potential issue (one that probably won't come up often but could).  Since if the poll succeeds then the while will call read.  I guess read is non blocking so would return the WouldBlock error (see comment below) but this could just spin until something showed up on the tty and then it would disappear in the void and this would fail (unless it happened to be a bel but at this point that is not likely).  I think you may need to incorporate the poll into the while condition (?) that way then while loop will not spin even in a degenerate case.  This might not handle the timeout 100% if something else came in via the tty other then what we want though...  Technically the timeout should be decremented by the time already waited OR maybe we should just remove the while loop and if the next thing in was not the bel then fail...  If other data is coming in on the tty then this could be an issue BUT I think this will consume and discard that data anyway so basically if you try to detect a color while the tty is writing stuff you are kinda hosed either way so maybe we just hose things in the simpler direction... 

Another wrinkle here, if conin has data queued before the escape code for this is written then there will be data already in the way (probably the most likely way to trip this up).  I think the only way to detect this currently would be to add a poll_timeout with a really small timeout before the write and return an error if data is ready to read (someone is being real sloppy...).  I kinda hate this though because this call will always be slowed down by the delay with potentially a mostly unneeded syscall to boot.  It may just be better to go with erroring out if it reads something other then the bel in the timeout (this probably means that someone called this with data queued in conin which is not good).  I think I like this better and hopefully a dev would see that error during dev and make their code not so shitty...

In short I think we should get rid of the while and just return an error if it reads something other then bel in the timeout but open to discussion.

File: src/cursor.rs [hunk 3]
```diff
@@ -179,37 +179,40 @@ impl<C: ConsoleRead> CursorPos for C {
...
-            // Either consume all data up to R or wait for a timeout.
-            while buf[0] != delimiter && now.elapsed().unwrap() < timeout {
-                match conin.read(&mut buf) {
-                    Ok(b) if b > 0 => read_chars.push(buf[0]),
-                    Ok(_) => {}
-                    // WouldBlock just means no data yet so keep trying.
-                    Err(err) if err.kind() == io::ErrorKind::WouldBlock => {}
-                    Err(err) => return Err(err),
+            if conin.poll_timeout(timeout) {
+                while buf[0] != delimiter {
```
Comment (sstanfield) [inline src/cursor.rs:184]: Same discussion as color here with basically same issues...

File: src/cursor.rs [hunk 4]
```diff
@@ -179,37 +179,40 @@ impl<C: ConsoleRead> CursorPos for C {
...
+                    match conin.read(&mut buf) {
+                        Ok(b) if b > 0 => read_chars.push(buf[0]),
+                        Ok(_) => {}
+                        // WouldBlock just means no data yet so keep trying.
+                        Err(err) if err.kind() == io::ErrorKind::WouldBlock => {}
```
Comment (sstanfield) [inline src/cursor.rs]: Can probably go away (see color note).

File: src/event.rs [hunk 5]
```diff
@@ -158,141 +174,215 @@ where
...
-            // (1, 1) are the coords for upper left.
-            let cx = next().saturating_sub(32) as u16;
-            let cy = next().saturating_sub(32) as u16;
-            Event::Mouse(match cb & 0b11 {
-                0 => {
-                    if cb & 0x40 != 0 {
-                        MouseEvent::Press(MouseButton::WheelUp, cx, cy)
-                    } else {
-                        MouseEvent::Press(MouseButton::Left, cx, cy)
+            if let (Some(cb), Some(cx), Some(cy)) =
```
Comment (sstanfield) [inline src/event.rs:207]: I like this, it is slick.

**Review by sstanfield — COMMENTED** [review]
The main issue I commented on is tricky but let me know what you think.  I think we are in a damned if you do and damned if you don't situation so should probably just make the code simpler.

Also, did you run the mouse example against this PR?  The tests are sketchy but the mouse example will exercise a lot of this stuff so you should run it if you have not (cargo run --example mouse).  There is a color detect example as well you should try.  You probably already did this, mentioning just in case.  I run the examples a lot since I think the current tests may be lacking due to the nature of this thing.

This is a great PR, thanks.
